packages/core/src/locales/en.ts

```diff
@@ -0,0 +1,127 @@
+import type { $ZodStringFormats } from "../checks.js";
+import type * as errors from "../errors.js";
+import * as util from "../util.js";
+
+const Sizable: Record<string, { unit: string; verb: string }> = {
+  string: { unit: "characters", verb: "to have" },
+  file: { unit: "bytes", verb: "to have" },
+  array: { unit: "items", verb: "to have" },
+  set: { unit: "items", verb: "to have" },
+};
+
+function getSizing(origin: string): { unit: string; verb: string } | null {
+  return Sizable[origin] ?? null;
+}
+
+export const parsedType = (data: any): string => {
+  const t = typeof data;
+
+  switch (t) {
+    case "number": {
+      return Number.isNaN(data) ? "NaN" : "number";
+    }
+    case "object": {
+      if (Array.isArray(data)) {
+        return "array";
+      }
+      if (data === null) {
+        return "null";
+      }
+
+      if (Object.getPrototypeOf(data) !== Object.prototype && data.constructor) {
+        return data.constructor.name;
+      }
+    }
+  }
+  return t;
+};
+
+const Nouns: {
+  [k in $ZodStringFormats | (string & {})]?: string;
+} = {
+  regex: "input",
+  email: "email address",
+  url: "URL",
+  emoji: "emoji",
+  uuid: "UUID",
+  uuidv4: "UUIDv4",
+  uuidv6: "UUIDv6",
+  nanoid: "nanoid",
+  guid: "GUID",
+  cuid: "cuid",
+  cuid2: "cuid2",
+  ulid: "ULID",
+  xid: "XID",
+  ksuid: "KSUID",
+  datetime: "ISO datetime",
+  date: "ISO date",
+  time: "ISO time",
+  duration: "ISO duration",
+  ipv4: "IPv4 address",
+  ipv6: "IPv6 address",
+  cidrv4: "IPv4 range",
+  cidrv6: "IPv6 range",
+  base64: "base64-encoded string",
+  base64url: "base64url-encoded string",
+  json_string: "JSON string",
+  e164: "E.164 number",
+  jwt: "JWT",
+  template_literal: "input",
+};
+
+const error: errors.$ZodErrorMap = (issue) => {
+  switch (issue.code) {
+    case "invalid_type":
+      return `Invalid input: expected ${issue.expected}, received ${parsedType(issue.input)}`;
+    // return `Invalid input: expected ${issue.expected}, received ${util.getParsedType(issue.input)}`;
+    case "invalid_value":
+      if (issue.values.length === 1) return `Invalid input: expected ${util.stringifyPrimitive(issue.values[0])}`;
+      return `Invalid option: expected one of ${util.joinValues(issue.values, "|")}`;
+    case "too_big": {
+      const adj = issue.inclusive ? "<=" : "<";
+      const sizing = getSizing(issue.origin);
+      if (sizing)
+        return `Too big: expected ${issue.origin ?? "value"} to have ${adj}${issue.maximum.toString()} ${sizing.unit ?? "elements"}`;
+      return `Too big: expected ${issue.origin ?? "value"} to be ${adj}${issue.maximum.toString()}`;
+    }
+    case "too_small": {
+      const adj = issue.inclusive ? ">=" : ">";
+      const sizing = getSizing(issue.origin);
+      if (sizing) {
+        return `Too small: expected ${issue.origin} to have ${adj}${issue.minimum.toString()} ${sizing.unit}`;
```

Comment:
```suggestion
        return `Too small: expected ${issue.origin ?? "value"} to have ${adj}${issue.minimum.toString()} ${sizing.unit ?? "elements"}`;
```

Reply:
Is there a scenario where `issue.origin` isn't being set properly? If so that's a bug. The `origin` property isn't optional so there's no need for a fallback here.